### PR TITLE
Make GetTensorShapeFromTensorShapeProto return TensorShape and not it's internal representation.

### PR DESCRIPTION
**Pull request by pranavsharma**
**Description**: Make GetTensorShapeFromTensorShapeProto return TensorShape and not it's internal representation.

**Motivation and Context**
GetTensorShapeFromTensorShapeProto says it returns TensorShape, but returns it's internal representation breaking encapsulation.

### CODE REVIEW

File: include/onnxruntime/core/framework/tensor_shape.h [hunk 1]
```diff
@@ -37,6 +37,7 @@ class TensorShape : private std::vector<int64_t> {
   TensorShape(const int64_t* dimension_sizes, size_t dimension_count);
 
   TensorShape(const std::vector<int64_t>& dims);
+  TensorShape(std::vector<int64_t>&& dims);
```
Comment (snnn) [inline include/onnxruntime/core/framework/tensor_shape.h]: It's not needed. std::vector<int64_t> can be cast into TensorShape  forcely.

Comment (snnn) [reply include/onnxruntime/core/framework/tensor_shape.h]: Please see the ReinterpretBaseType function in this file

File: onnxruntime/core/framework/tensorprotoutils.cc [hunk 2]
```diff
@@ -274,14 +274,14 @@ common::Status GetSizeInBytesFromTensorProto(const ONNX_NAMESPACE::TensorProto&
...
-std::vector<int64_t> GetTensorShapeFromTensorShapeProto(const ONNX_NAMESPACE::TensorShapeProto& tensor_shape_proto) {
+TensorShape GetTensorShapeFromTensorShapeProto(const ONNX_NAMESPACE::TensorShapeProto& tensor_shape_proto) {
   const auto& dims = tensor_shape_proto.dim();
   std::vector<int64_t> tensor_shape_vec(static_cast<size_t>(dims.size()));
   for (int i = 0; i < dims.size(); ++i) {
     tensor_shape_vec[i] = dims[i].has_dim_param() ? -1 /* symbolic dimensions are represented as -1 in onnxruntime*/
                                                   : dims[i].dim_value();
   }
-  return tensor_shape_vec;
+  return TensorShape(std::move(tensor_shape_vec));
```
[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/framework/tensorprotoutils.cc]: Not really. Casting assumes tensor_shape_vec is going to live beyond this function, which is not the case here.

File: include/onnxruntime/core/framework/tensor_shape.h (hunk 1 above)
Comment (pranavsharma) [reply include/onnxruntime/core/framework/tensor_shape.h]: see comment above.

File: onnxruntime/core/framework/tensorprotoutils.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (snnn) [reply onnxruntime/core/framework/tensorprotoutils.cc]: Then you can change tensor_shape_vec as a TensorProto, and you may still use it as a vector if you want. Casting is faster than move construction.

[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/framework/tensorprotoutils.cc]: I don't quite follow and would like to understand the performance difference here. You've to construct the vector in any case. The vector is getting constructed only once either way. The move constructor is simply copying pointers. Is this really slow? Can you quantify?

I don't think anyone should bother about casting if you return a TensorShape object as promised by the function. vector is an implementation detail of TensorShape.

File: onnxruntime/core/providers/cpu/controlflow/loop.cc [hunk 3]
```diff
@@ -422,7 +422,8 @@ Status LoopImpl::Execute(FeedsFetchesManager* ffm, const FeedsFetchesManager* ca
       if (graph_output_shape) {
         output_dims.reserve(graph_output_shape->dim_size() + 1);
 
-        auto dims = onnxruntime::utils::GetTensorShapeFromTensorShapeProto(*graph_output_shape);
+        const auto& tensor_shape = onnxruntime::utils::GetTensorShapeFromTensorShapeProto(*graph_output_shape);
+        const auto& dims = tensor_shape.GetDims();
```
Comment (linkerzhang) [inline onnxruntime/core/providers/cpu/controlflow/loop.cc]: Aha, when reading this line, I am wondering what's the motivation of this change please. The function name (GetTensorShapeFromTensorShapeProto) may not be good though :)

Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/controlflow/loop.cc]: The function name seems to be good enough. It just converts TensorShapeProto to TensorShape. Reasonable enough.
